vllm/utils.py

```diff
@@ -1363,6 +1364,37 @@ def value(self):
         return self._value
 
 
+@dataclass
+class BeamSearchSequence:
+    """A sequence for beam search.
+    It keeps track of the tokens and the log probability of the sequence.
+    The text field is optional and will only be filled when the sequence is
+    about to be returned to the user.
+    """
+    # The tokens includes the prompt.
+    tokens: List[int]
+    cum_logprob: float = 0.0
+    text: Optional[str] = None
+
+
+@dataclass
+class BeamSearchOutput:
+    """The output of beam search.
+    It contains the list of the best beam search sequences.
+    The length of the list is equal to the beam width.
+    """
+    sequences: List[BeamSearchSequence]
+
+
+class BeamSearchInstance:
+
+    def __init__(self, prompt_tokens: List[int]):
+        self.beams: List[BeamSearchSequence] = [
+            BeamSearchSequence(tokens=prompt_tokens)
+        ]
+        self.completed: List[BeamSearchSequence] = []
+
+
```

Comment:
There's a circular import error if I put these classes in `vllm/sequence.py` as `BeamSearchSequence` is needed in `vllm/utils.py`, but `vllm/sequence.py` indirectly imports from `vllm/utils.py`

Reply:
logically, `vllm/utils.py` should not import `vllm/sequence.py` . we should change the code if this is the case.

Reply:
should i also move the `create_sort_beams_key_function` to `vllm/sequence.py`?

Reply:
sure, go ahead!

Reply:
I'm not sure whether these should go in `sequence.py` since that holds "internal" data-structures used within the scheduler etc. and if I understand correctly, `BeamSearchSequence` etc. are only used in the outer layer(s). Maybe better to have a dedicated file an the appropriate place in the tree for this?

Reply:
> Maybe better to have a dedicated file an the appropriate place in the tree for this?

makes sense. how about `vllm/beam_search.py` ?